Revised local state docs

hwillson opened the pull request:
[Preview link](https://deploy-preview-4378--apollo-client-docs.netlify.com/docs/react/essentials/local-state.html).

----

This PR covers the Apollo Client local state management docs changes. It’s essentially a remix of the existing [Apollo Client local state docs](https://www.apollographql.com/docs/react/essentials/local-state.html) and  [`apollo-link-state` docs](https://www.apollographql.com/docs/link/links/state.html), merged with new feature content. 

A few notes:

- These docs aren’t intended to be a step by step tutorial, or line up with a reference or tutorial app. My thinking here was that since we’re already doing this with the [Apollo tutorial](https://www.apollographql.com/docs/tutorial/introduction.html) (which includes AC local state), we shouldn’t try to replicate things with another tutorial based approach (that we then have to keep up to date). These docs use code samples to explain things, but those code samples aren’t necessarily taken from a reference application (in some cases they are borrowed from the `fullstack-tutorial` app). If anyone disagrees with this approach, definitely let me know.

- I had a hard time deciding on section ordering. For example, should we explain how to manage the cache first (using `cache.writeData` for example), then how to use local resolvers? Or should resolvers be explained first, so people know why they might need to use something like `cache.writeData`? Should we show people how to use resolvers for mutations first so they have data in place before trying to query with a local resolver? Or should we explain querying with resolvers  first since it’s easier to grasp, before diving into mutations? I’ve gone with the approach implemented, but let me know if the ordering seems off - happy to re-work any of it!

- I think the section titles need work - feedback appreciated!

- Since knowing how to use certain `ApolloCache` methods is important when working with local state, I’m thinking it might make sense to add the  `ApolloCache` API to the left hand docs menu, down at the bottom in the API section (probably under `Apollo Client` / before `React Apollo`). Let me know if others agree.

- I think we should probably add an `Advanced` section, and in it give examples of doing things like loading resolvers on specific routes to better leverage code-splitting. Then again, maybe that’s overkill for the initial release (we can always add it in afterwards).

Thanks!

**Note:** I've left the base branch for this PR as `master` to use our Netlify docs preview. It won't work when the base branch is set to anything else, but when we're happy with the changes here, the base branch should be changed to `release-2.5.0`.

hwillson commented on docs/source/essentials/local-state.md:
```diff
@@ -215,118 +295,809 @@ const TodoList = () => (
+
+If you open up Apollo Client Devtools and click on the `GraphiQL` tab, you'll be able to explore your client schema in the "Docs" section. This example doesn't include a remote schema, but if it did, you would be able to see your local queries and mutations alongside your remote ones.
+
+![GraphiQL Console](../img/client-schema.png)
```
For some reason this image isn't showing properly in the Netlify preview. Hexo was corrupting the file, but when I moved it to a new directory and tested locally (with a fresh Hexo cache), that seems to have fixed the issue. I'm guessing the deploy preview is just in a bit of a broken state with regards to the image, but I'll look into this further.

Just to add - this image is also currently broken in production (https://www.apollographql.com/docs/react/essentials/local-state.html#schema). So Hexo has been corrupting the image for a while ... 🤔

rajington replied:
@hwillson I think it accidentally broke during the LF conversion, in this commit: https://github.com/apollographql/apollo-client/pull/4232/commits/53d33eb99a02722ff7edb8e9ddbc3cc72e750afc

I could make an undo PR for just the `png`s if you'd like.

rajington replied:
#4438

hwillson replied:
Oh, wow @rajington - awesome find! Thanks for the PR!

rajington replied:
absolutely! and great work on this project, glad I could help!

benjamn commented on docs/source/essentials/local-state.md:
```diff
@@ -215,118 +295,809 @@ const TodoList = () => (
+<h4 id="async-resolvers">Async local resolvers</h4>
 
-<h2 id="combine-data">Combining local and remote data</h2>
+Apollo Client supports asynchronous local resolver functions. These functions can either be `async` functions or ordinary functions that return a Promise. This can be useful for performing side effects like accessing a device API.
```
```suggestion
Apollo Client supports asynchronous local resolver functions. These functions can either be `async` functions or ordinary functions that return a `Promise`. Asynchronous resolvers are useful when they need to return data from an asynchronous API.
```

benjamn commented on docs/source/essentials/local-state.md:
```diff
@@ -215,118 +295,809 @@ const TodoList = () => (
+  cache: new InMemoryCache(),
+  resolvers: {
+    Query: {
+      cameraRoll: async (_, { assetType }) => {
```
```suggestion
      async cameraRoll(_, { assetType }) {
```

benjamn commented on docs/source/essentials/local-state.md:
```diff
@@ -215,118 +295,809 @@ const TodoList = () => (
-What’s really cool about using a `@client` directive to specify client-side only fields is that you can actually combine local and remote data in one query. Since we don't have a remote server, let's look at an example from [Pupstagram](https://codesandbox.io/s/r5qp83z0yq) where we add a client-only field to our remote data.
+> ⚠️ If you would like to hit a REST endpoint from your resolver, [we recommend checking out `apollo-link-rest`](https://github.com/apollographql/apollo-link-rest) instead, which is a more complete solution for using REST endpoints with Apollo Client.
+
+For React Native and most browser APIs, you should set up a listener in a component lifecycle method and pass in your mutation trigger function as the callback instead of using an async resolver. However, there are some cases where it's beneficial to perform a side effect within a resolver:
```
```suggestion
For React Native and most browser APIs, you should set up a listener in a component lifecycle method and pass in your mutation trigger function as the callback instead of using an async resolver. However, an `async` resolver function is often the most convenient way to consume asynchronous device APIs:
```

benjamn commented on docs/source/essentials/local-state.md:
````diff
@@ -215,118 +295,809 @@ const TodoList = () => (
+```
 
-First, let's look at an example of a mixed query. The `images` field comes from the server. When the array of images comes back, we add a local field `isLiked` to each image.
+[`CameraRoll.getPhotos()`](https://facebook.github.io/react-native/docs/cameraroll.html#getphotos) returns a Promise resolving to an object with a `edges` property, which is an array of camera node objects, and a `page_info` property, which is an object with pagination information. This is a great use case for GraphQL, since we can filter down the return value to only the data that our components ...
````
```suggestion
[`CameraRoll.getPhotos()`](https://facebook.github.io/react-native/docs/cameraroll.html#getphotos) returns a `Promise` resolving to an object with a `edges` property, which is an array of camera node objects, and a `page_info` property, which is an object with pagination information. This is a great use case for GraphQL, since we can filter down the return value to only the data that our components consume.
```

benjamn commented on docs/source/essentials/local-state.md:
```diff
@@ -215,118 +295,809 @@ const TodoList = () => (
-const resolvers = {
-  Image: {
-    isLiked: () => false
+As outlined in [Handling `@client` fields with resolvers](#client-fields-resolvers), `@client` fields can be resolved with the help of local resolver functions. While local resolvers can be super helpful, it's important to note that they are not always required when using an `@client` directive. Fields marked with `@client` can still be resolved locally, by pulling matching values out of the cach ...
```
```suggestion
As outlined in [Handling `@client` fields with resolvers](#client-fields-resolvers), `@client` fields can be resolved with the help of local resolver functions. However, it's important to note that local resolvers are not always required when using an `@client` directive. Fields marked with `@client` can still be resolved locally, by pulling matching values out of the cache directly. For example:
```

benjamn commented on docs/source/essentials/local-state.md:
```diff
@@ -215,118 +295,809 @@ const TodoList = () => (
+        const previous = cache.readQuery({ query });
+        const newTodo = { id: nextTodoId++, text, completed: false, __typename: 'TodoItem' };
+        const data = {
+          todos: previous.todos.concat([newTodo]),
```
```suggestion
          todos: [...previous.todos, newTodo],
```

benjamn commented on docs/source/essentials/local-state.md:
```diff
@@ -215,118 +295,809 @@ const TodoList = () => (
+
+The [`apollo-link-state`](https://github.com/apollographql/apollo-link-state) project was the first to bring local state handling into the Apollo ecosystem. Handling local resolvers through the addition of an `ApolloLink` was a great starting point, and proved that `@client` based queries make sense, and work really well for local state management.
+
+While `apollo-link-state` works well, and it's nice to be able to modularly include local state handling when required, the use of an `ApolloLink` introduces a few hard to work around limitations. Local state handling in a link is restricted to only being able to leverage the API elements available within the link. We see local state management as an important part of the Apollo ecosystem, and as ...
```
```suggestion
While `apollo-link-state` achieved some of the goals of local state handling, the information available when using any `ApolloLink` is limited by the modularity of the link system. We consider local state management a core part of the Apollo ecosystem, and as Apollo Client progresses, we want to make sure local resolvers are integrated as tightly as possible into core. This integration opens up new possibilities (like `@export` handling) and ties nicely into the future planned adjustments to cache data retention, invalidation, garbage collection, and other planned features that impact both local and remote data.
```

benjamn commented on docs/source/essentials/local-state.md:
````diff
@@ -215,118 +295,809 @@ const TodoList = () => (
+    resolvers: { ... },
+  });
+  ```
+  It's important to note here that `defaults` are no longer supported. To prep the cache, use [`cache.writeData`](#write-data) directly instead. So
````
Should this be a separate list item?

hwillson replied:
Yes, I think so - great point!

jangerhofer commented on docs/source/essentials/local-state.md:
```diff
@@ -215,118 +295,811 @@ const TodoList = () => (
+<h3 id="cache-initialization">Initializing the cache</h3>
 
-Your `defaults` object represents the initial state that you would like to write to the cache. It's important to provide defaults for your client-side queries, otherwise they could potentially error out if a mutation hasn't occurred before they run. The shape of your initial state should match how you plan to query it in your application.
+Often, you'll need to write an initial state to the cache so any components querying data before a mutation is triggered don't error out. To accomplish this, you can use `cache.writeData` to prep the cache with initial values. The shape of your initial state should match how you plan to query it in your application.
```
[This comment](https://github.com/apollographql/apollo-client/pull/4155#issuecomment-455174470) about async initializers cleared up one aspect of the new API which had caused me not-insignificant confusion.  Does the synchronous nature of constructor initializers bear mention here?

hwillson replied:
Thanks for the feedback @jangerhofer. We've decided to remove initializers from Apollo Client, and direct people towards using the existing Apollo Cache API to prepare the cache (`cache.writeData` specifically).

scaryguy commented:
Hi Apollo team,

Tonight, I had decided on diving deep into the issue that I wasn't able to get my head around since a while: Handling local (application) state within a React app that use Apollo Client. 

Previously, I remember that I had read that Apollo Client was internally using Redux in the official docs. Then, I learnt that that was the case at the first version but you no longer use Redux internally. After that I dived deep into [Apollo Link State docs ](https://www.apollographql.com/docs/link/links/state.html) to learn more about current implementation and how to use it. I read the [announcement post](https://blog.apollographql.com/the-future-of-state-management-dd410864cae2?_ga=2.90261658.1407696011.1549269012-1749799260.1543201592), watched the whole [Video tutorial by Sara Vieira](https://youtu.be/2RvRcnD8wHY). 

Finally, I've seen that link-state is being moved into core at the version 2.5 and the docs that I was supposed to be reading was [this](https://deploy-preview-4378--apollo-client-docs.netlify.com/docs/react/essentials/local-state.html). 

It was a little frustrating for me to waste a lot of time with out-dated docs (Especially the YouTube video. Using higher order components for event based mutations?). 

I think you'd better update all related docs/blog posts/links that I mentioned above accordingly or others that I didn't encounter. 

The first thing I'd do would be removing the first line from [here](https://www.apollographql.com/docs/link/links/state.html) and add a link navigating to [here](https://www.apollographql.com/docs/react/essentials/local-state.html). 

Thanks for all the effort and the good job.

hwillson commented:
Thanks @scaryguy - as soon as these docs are ready (early this week) and we launch Apollo Client 2.5.0 (later this week), we're going to update/deprecate all of the out of date local state docs, across the board. Sorry for the current headache, and thanks for the thoughtful feedback!

scaryguy commented:
> Thanks @scaryguy - as soon as these docs are ready (early this week) and we launch Apollo Client 2.5.0 (later this week), we're going to update/deprecate all of the out of date local state docs, across the board. Sorry for the current headache, and thanks for the thoughtful feedback!

Who knows how many people  in a week may have the bad experience that I've just had. I'd like to contribute by removing the first line from [here](https://github.com/apollographql/apollo-link-state/blob/master/README.md) and updating the issue # at the beginning of [this README](https://github.com/apollographql/apollo-link-state/blob/master/README.md) if you think that they'd help?

borisyordanov commented:
@hwillson 
Any updates on when 2.5 will launch?

Can you tell us which versions of apollo-client and react-apollo we should be using to implement the new local caching without waiting for 2.5?

[This ](https://deploy-preview-4378--apollo-client-docs.netlify.com/docs/react/essentials/local-state.html)is the most up to date documentation on local state management, right?

hwillson commented on docs/source/essentials/local-state.md:
```diff
@@ -1,58 +1,134 @@
+import { InMemoryCache } from 'apollo-cache-inmemory';
 
-If you'd like a deep dive into the `clientState` config properties, we recommend checking out the [`apollo-link-state` docs](/docs/link/links/state.html). Otherwise, get ready to learn about these properties gradually as we build `Query` and `Mutation` components for local data.
+cache.writeData({
```
Let's show it being created and keep the import. We've accepted a lot of AC docs PR's that add imports and other details, to make it more clear what's happening, so people definitely like this extra detail. Thanks!

hwillson commented on docs/source/essentials/local-state.md:
```diff
@@ -1,58 +1,134 @@
+  <dt>`setTypeDefs(typeDefs: string | string[] | DocumentNode | DocumentNode[])`</dt>
+  <dd>A string representing your client-side schema written in the [Schema Definition Language](/docs/graphql-tools/generate-schema.html#schema-language), or as a GraphQL AST. This schema is not used for validation, but is used for introspection by the [Apollo Client Devtools](https://github.com/apollographql/apollo-client-devtools).</dd>
+  <dt>`getTypeDefs`</dt>
+  <dd>Returns a string (SDL) or GraphQL AST representation of your client-side schema (if previously set via the `ApolloClient` constructor or `setTypeDefs`).</dd>
```
We've talked about this separately, and have decided to remove the `typeDefs` getters/setters. For history's sake, they were included to make it easier to get `typeDefs` in Apollo Client Devtools. We've since decided to let devtools access the private `typeDefs` member (for now, since devtools aren't using TS). Thanks!

hwillson commented on docs/source/essentials/local-state.md:
```diff
@@ -1,58 +1,134 @@
+<h2 id="updating-local-state">Updating local state</h2>
 
-There are two ways to perform mutations in `apollo-link-state`. The first way is directly writing to the cache by calling `cache.writeData` within an `ApolloConsumer` or `Query` component. Direct writes are great for one-off mutations that don't depend on the data that's currently in the cache, such as writing a single value. The second way is creating a `Mutation` component with a GraphQL mutati ...
+There are two main ways to perform local state mutations. The first way is to directly write to the cache by calling `cache.writeData` within an `ApolloConsumer` or through a `Query` component. Direct writes are great for one-off mutations that don't depend on the data that's currently in the cache, such as writing a single value. The second way is by creating a `Mutation` component with a GraphQ ...
```
Great revision - and great idea @peggyrayzis; I'll make an FR note of that!

hwillson commented on docs/source/essentials/local-state.md:
```diff
@@ -105,42 +181,48 @@ const FilterLink = ({ filter, children }) => (
+1. `obj`: The object containing the result returned from the resolver on the parent field or the `ROOT_QUERY` object in the case of a top-level query or mutation.
 2. `args`: An object containing all of the arguments passed into the field. For example, if you called a mutation with `updateNetworkStatus(isConnected: true)`, the `args` object would be `{ isConnected: true }`.
-3. `context`: The context object, which is shared between your React components and your Apollo Client network stack. The most important thing to note here is that we've added the Apollo cache to the context for you, so you can manipulate the cache with `readQuery`, `writeQuery`, `readFragment`, `writeFragment`, and `writeData`. Learn more about those methods [here](../advanced/caching.html#direc ...
+3. `context`: The context object, which is shared between your React components and your Apollo Client network stack. The most important thing to note here is that we've added the Apollo cache to the context for you, so you can manipulate the cache with `readQuery`, `writeQuery`, `readFragment`, `writeFragment`, and `writeData`. You can learn more about these methods in [Managing the cache](#mana ...
```
Great revision @benjamn, thanks! 

@peggyrayzis things can definitely get a bit ugly by allowing the full `client` instance to be available. We went back and forth on this, and decided to allow it since this feature was so highly requested. If people call `clearStore/resetStore` (and they don't handle resetting default values themselves), they can definitely cause problems. Good idea on the bulleted list, and maybe I'll add a small message here about being careful when calling parts of the `client` API.

hwillson commented on docs/source/essentials/local-state.md:
````diff
@@ -215,118 +295,809 @@ const TodoList = () => (
+```
 
-<h2 id="schema">Client-side schema</h2>
+Sometimes you may need to [reset the store](/docs/react/features/cache-updates.html#reset-store) in your application, when a user logs out for example. If you call `client.resetStore` anywhere in your application, you will likely want to initialize your cache again. You can do this using the `client.onResetStore` method to register a callback that will call `cache.writeData` again.
````
Great idea - I'll add a code snippet.

hwillson commented on docs/source/essentials/local-state.md:
```diff
@@ -1,58 +1,134 @@
+
+const client = new ApolloClient({
+  cache: new InMemoryCache(),
+  link: ApolloLink.empty(),
```
Good point - I'll switch it to a `HttpLink`. Thanks!

hwillson commented on docs/source/essentials/local-state.md:
```diff
@@ -1,58 +1,134 @@
+  <dd>A map of resolver functions that your GraphQL queries and mutations call in order to read and write to the cache. Resolver functions added through `setResolvers` overwrite all existing resolvers (a pre-existing resolver map is wiped out, before the new resolvers are added).</dd>
+  <dt>`getResolvers`</dt>
+  <dd>Get the currently defined resolver map.</dd>
+  <dt>`setTypeDefs(typeDefs: string | string[] | DocumentNode | DocumentNode[])`</dt>
```
We've removed the `typeDefs` getters/setters (to help avoid this confusion). Thanks!

hwillson commented on docs/source/essentials/local-state.md:
````diff
@@ -215,118 +295,811 @@ const TodoList = () => (
+Setting resolvers through `ApolloClient`'s constructor `resolvers` parameter, or through its `setResolvers` / `addResolvers` methods, adds resolvers to Apollo Client's internal resolver map (refer to the [Local resolvers](#local-resolvers) section for more details concerning the resolver map). In the above example we added a  `isInCart` resolver, for the `Launch` GraphQL object type, to the resol ...
 
+```js
+isInCart: (launch, _args, { cache }) => {
````
Great point and will do. For some reason I couldn't get line highlighting to work properly, so I left it out 🤔.

hwillson commented on docs/source/essentials/local-state.md:
```diff
@@ -215,118 +295,811 @@ const TodoList = () => (
+
+While `@client(always: true)` ensures that a local resolver is always fired, it's important to note that if a query is using a `fetchPolicy` that leverages the cache first (`cache-first`, `cache-and-network`, `cache-only`), the query is still attempted to be resolved from the cache first, before the local resolver is fired.    This happens because `@client(always: true)` use could be mixed with n ...
+
+<h2 id="combine-local-remote">Combining local and remote data</h2>
```
Hmmm - yes, it is a bit duplicative. I kind of like keeping the ideas a bit separate though, so we're showing that `@client` fields can be used without remote data, but I think we can merge these sections together. I'll get rid of the `Combining local and remote data` section completely, and move `Integrating @client into remote queries` up into the `Handling @client fields with resolvers` as a sub-section. I'll then move `Using @client fields as variables` up into the `Querying local state` section. Let's see how that looks. Thanks!

hwillson commented on docs/source/essentials/local-state.md:
```diff
@@ -215,118 +295,811 @@ const TodoList = () => (
+
+3. If you define multiple `@export` variables that use the same name, in a single operation, the value of the last `@export` variable will be used as the variable value moving forward. When this happens Apollo Client will log a warning message (dev only).
+
+<h2 id="managing-the-cache">Managing the cache</h2>
```
I’m on the fence on this one. I definitely hear what you’re saying, but working with the cache is such an integral part of the AC local state experience, that I wanted to make sure people could find out how to use it, without jumping around the docs. If we remove the `Managing the cache` section, and just point people to https://www.apollographql.com/docs/react/advanced/caching.html#direct, are we having them jump around a bit too much? Maybe not … one things for sure though, we’ll need to add `cache.writeData` to the `Direct Cache Access` section, since it isn’t mentioned anywhere.

So just to confirm @peggyrayzis, when you say point people to the advanced caching guide, you mean point them to [`Direct Cache Access`](https://www.apollographql.com/docs/react/advanced/caching.html#direct), correct? If so I’ll add `cache.writeData` to that section, remove `Managing the cache`, and update all links. Thanks!

hwillson commented on docs/source/essentials/local-state.md:
```diff
@@ -215,118 +295,811 @@ const TodoList = () => (
+
+To write the data to the cache, you can use either `cache.writeFragment` or `cache.writeData`. The only difference between the two is that `cache.writeFragment` requires that you pass in a fragment to validate that the shape of the data you're writing to the cache node is the same as the shape of the data required by the fragment. Under the hood, `cache.writeData` automatically constructs a fragm ...
+
+<h2 id="client-schema">Client-side schema</h2>
```
Hmm - good question. I'll think about this one a bit longer ...

hwillson commented on docs/source/essentials/local-state.md:
```diff
@@ -215,118 +295,811 @@ const TodoList = () => (
+
+const client = new ApolloClient({
+  cache: new InMemoryCache(),
+  link: ApolloLink.empty(),
```
For sure - thanks!

hwillson commented:
Thanks for the extremely thorough feedback @peggyrayzis! I've addressed most of your comments, but have a few questions/comments on the items in your main comment:

> Client-side schemas should have way more visibility, both in the docs and in the release post. All of the amazing tooling benefits that we promised a year ago are finally here, so we should show them off. No other state management solution gives you autocomplete + intellisense in VS Code and automatic generation of TypeScript definitions! 

It would be great to highlight this more thoroughly, but does this belong in the local state docs, or would it be more at home in our tooling docs? I'm guessing this should probably live in our tooling docs, and I would then add something to the client schema section of these docs, mentioning that client side schemas can feed into our tooling, and point people to the tooling docs.

> Another cool use case that @debergalis came up with is mocking on the client -- before a product team builds a subsection of the schema into their API, they can easily mock it by creating a client schema and resolvers to contextually mock out their data. When they're ready to integrate it into the remote API, they can drop the client directives. Several customers have been really impressed by that suggestion, so we should include them in the docs.

I believe this is now going to be addressed separately (and more thoroughly). Should we leave this one for now, and update the docs after launch, when the other local mocking initiaive is in place?

> What about code splitting? We mention the methods for dynamically adding resolvers in the API section but don't actually provide a use case. Now that initializers are gone, I'd also like to see an example of how to write values to the cache from a dynamically loaded component.

Sounds good - I'll add this in.

> [Cache persistence](https://github.com/apollographql/apollo-link-state/issues/170) is another initializer use case that just came to mind.

AC 2.5 should still be able to work with cache persistence without initializers. People can just call `cache.writeData` directly (and synchronously) before calling `persistCache`.

> The docs are thorough, but they feel a bit long for an essentials guide. I think moving the cache methods out of this guide should help a bit -- do you think there's anything else that could be moved or cut?

I agree, but I'm at a loss for what else should be cut out 🙁. Definitely open to more suggestions!

> Are subscriptions a thing now? I don't know how useful they will be, but we should include an example if they work!

They are! I'll get an example in.

> Versioning the docs: Do we want to version them for this release? We have versioning built into the Apollo Server docs and the new Gatsby infra.

Good point - we'll want versioning. I'll look into getting this wired up.

dko-slapdash commented:
setTypeDefs is mentioned at https://www.apollographql.com/docs/react/essentials/local-state/, although this method doesn't exist anymore
